compiler/rustc_const_eval/src/transform/check_consts/check.rs

```diff
@@ -355,10 +355,19 @@ impl<'mir, 'tcx> Checker<'mir, 'tcx> {
             // to mutable memory.
             hir::ConstContext::ConstFn => self.check_op(ops::TransientMutBorrow(kind)),
             _ => {
+                // For indirect places, we are not creating a new permanent borrow, it's
+                // just as transient as the already existing one. For reborrowing references
+                // this is handled at the top of `visit_rvalue`, but for raw pointers we handle it
+                // here. Pointers/references to `static mut` also end up here.
                 // Locals with StorageDead do not live beyond the evaluation and can
                 // thus safely be borrowed without being able to be leaked to the final
                 // value of the constant.
-                if self.local_has_storage_dead(local) {
+                // Note: This is only sound if every local that has a `StorageDead` has a
+                // `StorageDead` in every control flow path leading to a `return` terminator.
+                // The good news is that interning will detect if any unexpected mutable
+                // pointer slips through.
+                if place.is_indirect_first_projection() || self.local_has_storage_dead(place.local)
```

Comment:
I assume this runs after the deref'er? Debug assertions would fire if the `Deref` was later in the projection chain.

Reply:
Ah, I get a bunch of debug assertion failures.^^

Hm, but... the "reborrow" checks only check the first projection. How does that make sense then?

Reply:
derefer is still after borrowck. We haven't managed to move it earlier yet

Reply:
This means `&mut (*ptr).field` will not hit the reborrow code path in const-checking. I wonder if there's an example here where this PR changes behavior because of that...

Reply:
Okay, I think for mutable references that reborrow logic largely serves to avoid duplicate errors, and to paper over particular desugarings (like how `_ as *mut _` becomes a reborrow).

I can't wait to rip out all this half-effective logic when we stabilize `const_mut_refs`.^^